Add compact block filter client via Kyoto

rustaceanrob opened the pull request:
### Description

Adds a compact block filter backend via Kyoto.

### Notes to the reviewers

This description has been updated to the latest forced push as of Jan 29

#### Architectural notes:

The language semantics of Rust make it difficult to access the same type in two different place within a project. Due to this the `bdk_kyoto` "client" is separated into smaller component types that handle logs, warnings, and wallet updates. Because the target languages for the bindings are garbage collected/reference counted, I think it is an advantage to unite all of these parts into a single `Client` type. When a user builds a new node, they receive a `LightNode` and `Client`. 

The `LightNode` must be run to do anything, and a user would do so with `LightNode.run()`.

The `Client` is responsible for everything else:
- Getting log messages with `next_log`
- Getting warnings
- Getting wallet `Update`
- Asking for the minimum fee rate to broadcast a transaction
- Adding new scripts
- Broadcasting transactions

Taking into account previous conversations, this was conveyed as a more intuitive approach than using callbacks to handle events.

Some additions in particular:
- The client can add all revealed scripts with `add_revealed_scripts` that uses a `&Wallet`. This would be called after a user creates a transaction or reveals a receiving address
- There is a simple `is_running` function to check on the node
- Developers must pick between three different `ScanType`. This is to prevent dangerous guess-work when building the client. When creating a new wallet, all that should occur is a quick sync to add the tip of the blockchain and addition of this block hash to the wallet database. This is accomplished with `ScanType::New`. Recoveries are handled with `ScanType::Recovery` where the user can select a particular height to scan from. Lastly, syncing from the last known block hash by the wallet is accomplished with `ScanType::Sync`, which is the default.

### Changelog notice

<!-- Notice the release manager should include in the release tag message changelog -->
<!-- See https://keepachangelog.com/en/1.0.0/ for examples -->

- Add a compact block filter chain source

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [ ] I've added docs for the new feature

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [ ] I'm linking the issue being fixed by this PR

thunderbiscuit commented:
Just for fun:

Binary size when including Kyoto, building the aarch64 binary on macOS using the release-smaller profile (libbdkffi.dylib): 9.7MB
Binary size without Kyoto, building the aarch64 binary on macOS using the release-smaller profile (libbdkffi.dylib): 8.6MB

Not bad at all for what is likely to be one of the most popular clients one day.

rustaceanrob commented:
Awesome thank you for checking that out. Props to `rust-bitcoin` and `tokio` for keeping things tidy.

rustaceanrob commented:
> Objects cannot currently be used in enum variant data

You can imagine my disappointment with this error message after hacking on this PR again. I'm trying to express the new `Event<K>` from `bdk_kyoto` but it returns objects like `Update` (`FullScanResult`) as a variant. We can return more primitive types as variants, but then we can't express the `Update`. It doesn't make sense to express an event as anything else than an enum, as only one variant is used at a time to express the event.

It's been so long but maybe that is what I had in mind with the callback interface in the first place lol. Open to new ideas, but until we can return objects from enum variants I think we may have to stick with callbacks...

rustaceanrob commented:
I consider this ready for review now. Some updates, I added live tests for android, JVM, swift, and python. Users now have to build a `Peer` with an `IpAddress` if they have a company or personal node they want to connect to. `NodeMessageHandler` is renamed to `NodeEventHandler`, which reflects a bit more what is happening. Some open questions I have:

Do we really need a separate live test in android when the JVM one exists? It's basically the exact same code and it would likely just increase the maintenance burden for no real benefit IMO. 

Should I add docstrings? I think so, but I wanted to confirm

thunderbiscuit commented:
To answer your questions @rustaceanrob:

1. No I don't think we need a live test on Android if we have one on JVM (moreso because the Android emulators don't currently work in the CI!)
2. Yes let's add docstrings.

Awesome! I'll try to review this again next week.

thunderbiscuit commented on bdk-jvm/lib/build.gradle.kts:
```diff
@@ -61,12 +62,14 @@ tasks.withType<Test> {
     // testImplementation("org.junit.jupiter:junit-jupiter-api:5.10.1")
     // testRuntimeOnly("org.junit.jupiter:junit-jupiter-engine:5.10.1")
     // testRuntimeOnly("org.junit.vintage:junit-vintage-engine:5.8.2")
+    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.8.1")
```
You don't need this import twice.

thunderbiscuit commented on bdk-jvm/lib/build.gradle.kts:
```diff
@@ -61,12 +62,14 @@ tasks.withType<Test> {
 dependencies {
     implementation(platform("org.jetbrains.kotlin:kotlin-bom"))
     implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk7")
+    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.8.1")
```
The coroutines library import is required on the Android side as well (`bdk-android/lib/build.gradle.kts`).

rustaceanrob replied:
Should be added

thunderbiscuit reviewed (commented):
Just writing down some stuff as I build and test.

thunderbiscuit commented:
I got it working on the Android example wallet! So far so good. I'll take a closer look at the code before giving the final ACK.

My first little API-related comment is that the `NodeEventHandler` doesn't have an unique event for a new block being mined but that's a significant event from the point of view of the user, particularly because they loose mempool information when using Kyoto (if blocks pass and the transaction they expect doesn't get mined they need to look into it).

At the moment this is what I do to pull the new blocks out of the events:

```kotlin
override fun dialog(dialog: String) {
    if (dialog.contains("peer height")) {
        val height = dialog.split("peer height: ")[1].split(" ")[0]
        triggerSnackbar("New block mined: $height")
    }
}
```

Which is basically a custom parsing of the string just to get what I need. I wonder if a `newBlock()` method on the interface would be cleaner.

rustaceanrob commented:
The best known peer height is found when connecting to a peer, but our height may not yet match theirs yet. The `Synced` event is emitted every time we sync to their height, which includes when they advertise a new block to us. In effect, I believe the `Synced` event is more accurate as to what height the client is synced to. However, I did add a recent `Progress` event that does contain the best known height of remote peers, so the user has an idea of how many filters and filter headers must be downloaded until synced. I will add that in the next `kyoto` and `bdk_kyoto` releases and update this PR along with some other improvements.

thunderbiscuit commented:
Oh sweet yes I had not looked at all the methods carefully enough. It was easy to update the app to use that instead. 👍

thunderbiscuit commented:
@rustaceanrob would you mind rebasing this? I think we're getting close to merging, or at least this PR is the next one in line for me.

rustaceanrob commented:
> would you mind rebasing this?

I've been continually rebasing I think this should be up to date

rustaceanrob commented on bdk-ffi/src/bdk.udl:
```diff
@@ -1,4 +1,10 @@
+  [Throws=LightClientBuilderError]
+  NodePair build_light_client([ByRef] Wallet wallet, sequence<Peer> peers, u8 connections, u32? recovery_height, string data_dir);
+  /// Run a [`LightNode`] continuously on a detached OS thread.
+  void run_node(LightNode node);
```
I think `run_node` is probably the better option for most, as it spawns a new OS thread and runs the node there, then immediately returns. I would expect this would be the workflow for the majority of use cases that just want to run it and forget about it, and it handles the sometimes complicated semantics of spawning threads in languages like python. That being said I think we should also keep `LightNode.run` in the public API because languages like swift that have the concept of a `Task` with different "priorities" might be able to force the node to stay alive even when the app is idle for instance.

rustaceanrob replied:
Okay. As we uncovered offline, we cannot expose `LightNode.run`, and will have to stick with `run_node`

rustaceanrob replied:
Would it make sense to move this to `Client::run_node` or is that non-intuitive?

rustaceanrob commented on bdk-ffi/src/bdk.udl:
```diff
@@ -1,4 +1,10 @@
-namespace bdk {};
+namespace bdk {
```
For the entire PR? Sounds good I can go with that

rustaceanrob commented on bdk-ffi/src/bdk.udl:
```diff
@@ -1086,6 +1106,166 @@ dictionary KeychainAndIndex {
+  void broadcast([ByRef] Transaction transaction);
+  /// Add another [`Address`] to the [`LightNode`] to watch for inclusion in new blocks.
+  [Async, Throws=LightClientError]
+  void watch_address(Address address);
```
1. Yes. Kyoto uses the `Wallet` `lookahead` to peek scripts into the future. This is of course set by the user but has a reasonable default of 30 I believe. 
2. The reason `watch_address` has to exist is the user may reveal more than the `lookahead` amount of scripts to receive a payment. Maybe not a casual user, but a donation wallet perhaps. I think the appropriate documentation is to say it is necessary to call `watch_address` after revealing the `lookahead` number of scripts, but there is no cost to calling it for every revealed script so the developer might as well do so.

thunderbiscuit commented on .github/workflows/test-swift.yaml:
```diff
@@ -24,4 +24,4 @@ jobs:
       - name: "Run Swift tests"
         working-directory: bdk-swift
-        run: swift test --skip LiveElectrumClientTests --skip LiveMemoryWalletTests --skip LiveTransactionTests --skip LiveTxBuilderTests --skip LiveWalletTests
+        run: swift test --skip LiveElectrumClientTests --skip LiveMemoryWalletTests --skip LiveTransactionTests --skip LiveTxBuilderTests --skip LiveWalletTests --skip LiveKyotoTests
```
The test is called LiveKyotoTest instead of LiveKyotoTests

thunderbiscuit replied:
Sorry that's the Swift tests! Disregard my comment.

rustaceanrob commented:
Clean up from the call:
- [ ] Add the fee filter so users can broadcast low-fee transactions without rejection
- [ ] Add the optional reject reason when `tx_failed`
- [ ] Convert `build_light_client` to builder pattern (avoid breaking changes in the future)

rustaceanrob commented:
Rebased and updated to `beta-6` in `bdk_kyoto`. 

Also introduced, `bdk_kyoto` now allows for a user to pass a `&Wallet` to add all revealed scripts to the node on behalf of the user. This would simultaneously resolve adding scripts for change and receiving, because the change script is revealed during transaction construction and when calling `reveal_next_address`. The new methods would on the event publisher would be `add_revealed_scripts(&wallet)` and `broadcast_tx(&tx)`. 

I think that is a preferable and unambiguous API, but will still get feedback before I change this PR

thunderbiscuit commented:
Small note: the Android library is missing a dependency on the `kotlinx-coroutine-core` library.

```kotlin
implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.8.1")
```

rustaceanrob commented:
Added to `bdk-android/lib/build-gradle.kts`. Not sure if that is correct or not

thunderbiscuit commented:
With the cut of the [release/1.1]() branch, I don't think we'll be merging anything else that'll make it in before the big release. If we end up needing to add anything to the release, we'll merge those commits right on the release branch and cherry-pick on master if needed.

This means master is free to keep going at its own pace and I think this PR is ready to merge (again it will not make it into the 1.1 but rather in 1.2).

No real rush, but also I feel like @rustaceanrob has been really patient and rebasing every few days... I have this working well in my `variant/kyoto` app (which needs a big rebase on the esplora variant which has added a bunch of cleanups and features, but works well nonetheless!). 

@reez what do you say? Have you had a chance to test it out a bit?
@rustaceanrob is this a version good to merge? Or are you expecting newer/better features in the next few days/weeks and would rather polish those up first?

rustaceanrob commented:
I can rebase this soon and I think it's ready to go. I would prefer to do follow-up patches for anything additional

thunderbiscuit reviewed (approved):
ACK d154117a125eec9fd3bee76db158e8722e5d2178.

reez commented:
ACK d154117a125eec9fd3bee76db158e8722e5d2178.
